tools/clang/tools/dxcompiler/dxcshadersourceinfo.h

```diff
@@ -0,0 +1,70 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// dxcshadersourceinfo.h                                                     //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Utility helpers for dealing with DXIL part related to shader sources      //
+// and options.                                                              //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include <vector>
+#include <stdint.h>
+#include "llvm/ADT/StringRef.h"
+#include "dxc/DxilContainer/DxilContainer.h"
+
+namespace clang {
+  class CodeGenOptions;
+  class SourceManager;
+}
+
+namespace hlsl {
+
+struct SourceInfo {
+  struct Source {
+    llvm::StringRef Path;
+    llvm::StringRef Content;
+  };
+
+  llvm::StringRef MainFile;
+  std::vector<llvm::StringRef> Args;
+  std::vector<llvm::StringRef> Defines;
+  std::vector<Source> Sources;
+};
+
+struct SourceInfoReader {
+  using Buffer = std::vector<uint8_t>;
+  Buffer m_UncompressedSources;
+
+  struct Source {
+    llvm::StringRef Name;
+    llvm::StringRef Content;
+  };
+
+  std::vector<Source> m_Sources;
+  llvm::StringRef m_Defines;
+  llvm::StringRef m_Args;
+  llvm::StringRef m_TargetProfile;
+  llvm::StringRef m_EntryPoint;
+
+  llvm::StringRef GetArgs() const;
+  llvm::StringRef GetDefines() const;
+  llvm::StringRef GetTargetProfile() const { return m_TargetProfile; }
+  llvm::StringRef GetEntryPoint() const { return m_EntryPoint; }
+  Source GetSource(unsigned i) const;
+  unsigned GetSourcesCount() const;
+  void Read(const hlsl::DxilSourceInfo *SourceInfo);
```

Comment:
It probably should be made explicitly clear that the memory pointed to by SourceInfo here (via StringRef objects) is being referenced by the data members of the SourceInfoReader after the call to Read completes (SourceInfoReader doesn't own the memory).  That's not what I naturally expect from a Read() method deserializing data.

Reply:
Changed the name to Init instead, and added a comment.